common/scala/src/main/scala/org/apache/openwhisk/core/database/KafkaActivationStore.scala

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.openwhisk.core.database
+
+import akka.actor.ActorSystem
+import akka.stream.ActorMaterializer
+import com.google.common.base.Throwables
+import org.apache.openwhisk.common.{Logging, TransactionId}
+import org.apache.openwhisk.core.{ConfigKeys, WhiskConfig}
+import org.apache.openwhisk.core.connector.{MessageProducer, MessagingProvider, ResultMessage}
+import org.apache.openwhisk.core.entity.{ActivationEntityLimit, ByteSize, DocInfo, WhiskActivation}
+import org.apache.openwhisk.spi.SpiLoader
+import org.apache.openwhisk.core.entity.size._
+import pureconfig.loadConfigOrThrow
+
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+case class KafkaActivationStoreConfig(activationsTopic: String, db: Boolean, maxRequestSize: Option[ByteSize])
+
+class KafkaActivationStore(producer: MessageProducer,
+                           config: KafkaActivationStoreConfig,
+                           actorSystem: ActorSystem,
+                           actorMaterializer: ActorMaterializer,
+                           logging: Logging)
+    extends ArtifactActivationStore(actorSystem, actorMaterializer, logging) {
```

Comment:
Is there any possibility that `KafkaActivationStore` can be used with other `ActivationStore` rather than just `ArtifactActivationStore`?

Reply:
I think it's hard for `KafkaActivationStore` to be compatible with other `ActivationStore` as different `ActivationStore` may have different definition for methods `store`, `get`, `delete` and `countActivationsInNamespace`, and so on, and we cannot define these methods except the `store` in the `KafkaActivationStore`

Reply:
Current ActivationPersisterService is also using the ActivationStore for storage. So in theory we can support others. 

I can introduce a config to capture base activation store and then have KafkaActivationStore act as decorator for that. Would give it a try

Reply:
This is implemented now. There is now a config to configure the primary store via `primary-store-provider` which defaults to `ArtifactActivationStoreProvider`

```hocon
whisk {
  kafka-activation-store {
    # Name of the Kafka topic for sending WhiskActivations
    # Set to
    # - `completed-others` - When LogDriverLogStore is used
    # - `activations` - When other LogStore is used where logs are stored as part of activation result
    activations-topic = "completed-others"

    # Primary activation store used for eventual storage and querying of activation records
    primary-store-provider = org.apache.openwhisk.core.database.ArtifactActivationStoreProvider

    # Also store directly to primary ActivationStore. This mode can be used for initial
    # trial runs where data is sent to both primary store and also via Kafka to ActivationPersisterService
    store-in-primary = false

    # Maximum request size. By default it uses the MAX_ACTIVATION_LIMIT as computed from
    # `whisk.activation.payload`. Bump it to higher value if the activation result also
    # includes the logs
    # max-request-size = 2 MB
  }
}
```

So one can configure any `ActivationStore` impl and `KafkaActivationStore` would wrap it and handles the store flow. Rest of the flow is directed back to primary store